services/cd-service/pkg/repository/transformer.go

```diff
@@ -1620,6 +1624,108 @@ func generateReleaseTrainResponse(envDeployedMsg, envSkippedMsg map[string]strin
+					return currentEnv
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func getOverrideVersions(commitHash, upstreamEnvName string, repo Repository) (resp []Overview, err error) {
+	oid, err := git.NewOid(commitHash)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating new oid for commitHash %s: %w", commitHash, err)
+	}
+	s, err := repo.StateAt(oid)
+	if err != nil {
+		var gerr *git.GitError
+		if errors.As(err, &gerr) {
+			if gerr.Code == git.ErrNotFound {
+				return nil, fmt.Errorf("ErrNotFound: %w", err)
+			}
+		}
+		return nil, fmt.Errorf("unable to get oid: %w", err)
+	}
+
+	if envs, err := s.GetEnvironmentConfigs(); err != nil {
+		return nil, fmt.Errorf("unable to get EnvironmentConfigs for %s: %w", commitHash, err)
+	} else {
+		result := mapper.MapEnvironmentsToGroups(envs)
+		for envName, config := range envs {
+			var groupName = mapper.DeriveGroupName(config, envName)
+			var envInGroup = getEnvironmentInGroup(result, groupName, envName)
+			if upstreamEnvName != envInGroup.Name || upstreamEnvName != groupName {
+				continue
+			}
+			if apps, err := s.GetEnvironmentApplications(envName); err != nil {
+				return nil, fmt.Errorf("unable to get EnvironmentApplication for env %s: %w", envName, err)
+			} else {
+				for _, appName := range apps {
+					app := api.Environment_Application{
+						Name: appName,
+					}
+					var version *uint64
+					if version, err = s.GetEnvironmentApplicationVersion(envName, appName); err != nil && !errors.Is(err, os.ErrNotExist) {
+						return nil, fmt.Errorf("unable to get EnvironmentApplicationVersion for %s: %w", appName, err)
+					} else {
+						if version == nil {
+							continue
+						} else {
+							app.Version = *version
+						}
+					}
+					resp = append(resp, Overview{App: app.Name, Version: app.Version})
+				}
+			}
+		}
+	}
+	return resp, nil
+}
+
+func (c *ReleaseTrain) getUpstreamLatestApp(upstreamLatest bool, state *State, ctx context.Context, upstreamEnvName, source, commitHash string) (apps []string, appVersions []Overview, err error) {
+	if commitHash != "" {
+		appVersions, err := getOverrideVersions(c.CommitHash, upstreamEnvName, c.Repo)
+		if err != nil {
+			return nil, nil, grpc.PublicError(ctx, fmt.Errorf("could not get app version for commitHash %s for %s: %w", c.CommitHash, c.Target, err))
+		}
+		// check that commit hash is not older than 20 commits in the past
+		for _, app := range appVersions {
+			apps = append(apps, app.App)
+			versions, err := findOldApplicationVersions(state, app.App)
+			if err != nil {
+				return nil, nil, grpc.PublicError(ctx, fmt.Errorf("unable to find findOldApplicationVersions for app %s: %w", app.App, err))
+			}
+			if len(versions) > 0 && versions[0] > app.Version {
+				return nil, nil, grpc.PublicError(ctx, fmt.Errorf("Version for app %s is older than 20 commits when running release train to commitHash %s: %w", app.App, c.CommitHash, err))
+			}
+
+		}
+		return apps, appVersions, nil
+	} else if upstreamLatest {
```

Comment:
nitpick: no need for an `else` here:
```suggestion
	}
	if upstreamLatest {
```